tests/averaging/rotation/test_cycle_consistency.py

```diff
@@ -0,0 +1,327 @@
+    """
+    i2Ri1_dict = {
+        (0, 1): Rot3(),
+        (1, 2): Rot3(),
+        (0, 2): Rot3(),
+        (2, 3): Rot3(),
+        (3, 4): Rot3(),
+        (2, 4): Rot3.Ry(np.deg2rad(15)),
+    }
+    i2Ui1_dict = {
+        (0, 1): Unit3(np.array([1, 0, 0])),
+        (1, 2): Unit3(np.array([1, 0, 0])),
+        (0, 2): Unit3(np.array([1, 0, 0])),
+        (2, 3): Unit3(np.array([1, 0, 0])),
+        (3, 4): Unit3(np.array([1, 0, 0])),
+        (2, 4): Unit3(np.array([1, 0, 0])),
+    }
+    # assume no ground truth information available at runtime
+    two_view_reports_dict = {}
+
+    for (i1, i2) in i2Ri1_dict.keys():
+        two_view_reports_dict[(i1, i2)] = TwoViewEstimationReport(
+            v_corr_idxs=np.array([]),  # dummy array
+            num_inliers_est_model=10,  # dummy value
+        )
+
+    i2Ri1_dict_consistent, i2Ui1_dict_consistent = cycle_utils.filter_to_cycle_consistent_edges(
+        i2Ri1_dict, i2Ui1_dict, two_view_reports_dict, visualize=True
+    )
+    # non-self-consistent triplet should have been removed
+    expected_keys = {(0, 1), (1, 2), (0, 2)}
+    assert set(i2Ri1_dict_consistent.keys()) == expected_keys
+
+
+def extract_triplets_brute_force(i2Ri1_dict: Dict[Tuple[int, int], Rot3]) -> List[Tuple[int, int, int]]:
+    """Use triple for-loop to find triplets from a graph G=(V,E) in O(n^3) time.
+
+    **Much** slower implementation for large graphs, when compared to `extract_triplets()` that uses intersection of adjacency lists.
+    Used to check correctness inside the unit test.
+
+    Args:
+        i2Ri1_dict: mapping from image pair indices to relative rotation.
+
+    Returns:
+        triplets: 3-tuples of nodes that form a cycle. Nodes of each triplet are provided in sorted order.
+    """
+    triplets = set()
+
+    for (i1, i2), i2Ri1 in i2Ri1_dict.items():
+        if i2Ri1 is None:
+            continue
+
+        for (j1, j2), j2Rj1 in i2Ri1_dict.items():
+            if j2Rj1 is None:
+                continue
+
+            for (k1, k2), k2Rk1 in i2Ri1_dict.items():
+                if k2Rk1 is None:
+                    continue
+
+                # check how many nodes are spanned by these 3 edges
+                cycle_nodes = set([i1, i2]).union(set([j1, j2])).union(set([k1, k2]))
+                # sort them in increasing order
+                cycle_nodes = tuple(sorted(cycle_nodes))
+
+                # nodes cannot be repeated
+                unique_edges = set([(i1, i2), (j1, j2), (k1, k2)])
+                edges_are_unique = len(unique_edges) == 3
+
+                if len(cycle_nodes) == 3 and edges_are_unique:
+                    triplets.add(cycle_nodes)
+
+    return list(triplets)
+
+
+def test_triplet_extraction_correctness_runtime() -> None:
+    """Ensure that for large graphs, the adjacency-list-based algorithm is faster and still correct,
+    when compared with the brute-force O(n^3) implementation.
+    """
```

Comment:
The time aspect can create false alarms. I am a bit wary of checking this in.

Reply:
The brute force method is unbelievably slow, with cubic complexity vs linear for the release implementation. Were you concerned about it being faster than the other impl and raising an assert?

Reply:
For some reason yeah. I think we do not need to enforce it via a test. But keep it as a note to future developers.